Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp

```diff
@@ -742,6 +743,17 @@ void WebLocalFrameLoaderClient::dispatchDidFinishLoad()
         loadListener->didFinishLoad(m_frame.ptr());
 
     webPage->didFinishLoad(m_frame);
+
+    if (!m_frame->coreLocalFrame() || !m_frame->coreLocalFrame()->isMainFrame() || !m_frame->coreLocalFrame()->document())
+        return;
+    auto* document = m_frame->coreLocalFrame()->document();
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::StorageAccessQuirkForTopFrameDomain(RegistrableDomain { document->url() }), [document, weakDocument = WeakPtr { *document }](Vector<RegistrableDomain>&& domains) {
```

Comment:
Could we just capture weakDocument = WeakPtr { document } without also capturing the raw pointer to the document and without dereferencing an unchecked but probably non-null pointer?

Also, is there a way we can do a quick check in the web content process to determine whether we need to send this message?  I imagine the small minority of frame loads will need this.

This could also be a separate FrameLoaderClient call because it doesn't really have anything to do with the finishing of the frame load.

Reply:
We already restrict this to the main frame. We can further restrict it to main frames where we have quirks. I'll pre-load that set into the web process.

I can move it this call into a separate method, the important timing is that this is called after the document is loaded and before any iframes are created.